app/src/main/java/ch/onepass/onepass/utils/TimeProvider.kt

```
@@ -0,0 +1,97 @@
+import com.google.firebase.Timestamp
+import com.google.firebase.functions.FirebaseFunctions
+import java.util.Date
+import javax.inject.Inject
+import javax.inject.Singleton
+import kotlinx.coroutines.tasks.await
+
+/**
+ * Universal interface for providing the current time. Using this interface instead of
+ * System.currentTimeMillis() allows for:
+ * 1. Synchronizing with server time to prevent tampering.
+ * 2. Mocking time in unit tests.
+ */
+interface TimeProvider {
+  /** Returns the current trusted timestamp. */
+  fun now(): Timestamp
+
+  /** Returns the current trusted Date. */
+  fun currentDate(): Date
+
+  /** Synchronizes the local clock offset with the server. Should be called on app startup. */
+  suspend fun syncWithServer()
+}
+
+/**
+ * Implementation that calculates the offset between device time and server time. If sync fails, it
+ * gracefully falls back to device time.
+ */
+@Singleton
+class ServerSyncedTimeProvider @Inject constructor(private val functions: FirebaseFunctions) :
+    TimeProvider {
+
+  // The difference in milliseconds: ServerTime - DeviceTime
+  private var clockOffset: Long = 0
+  private var isSynced: Boolean = false
+
+  override fun now(): Timestamp {
+    val adjustedTimeMillis = System.currentTimeMillis() + clockOffset
+    return Timestamp(Date(adjustedTimeMillis))
+  }
+
+  override fun currentDate(): Date {
+    return now().toDate()
+  }
+
+  override suspend fun syncWithServer() {
+    try {
+      val startTime = SystemClock.elapsedRealtime()
+
+      // Call the Cloud Function
+      val result = functions.getHttpsCallable("getServerTime").call().await()
+
+      val endTime = SystemClock.elapsedRealtime()
+      val latency = (endTime - startTime) / 2 // Approximate one-way latency
+
+      val data = result.data as? Map<String, Any>
+      val serverTimeMillis =
+          (data?.get("timestamp") as? Number)?.toLong()
+              ?: throw IllegalStateException("Invalid server response")
```

The current Firebase response parsing in `ServerSyncedTimeProvider` uses a cast to `Map<String, Any>`:

```kotlin
val data = result.data as? Map<String, Any>
````

This approach is brittle because it can fail if the structure of the response changes or if the `timestamp` field is returned as a `Double` or `Long` instead of the expected type. Such issues could lead to runtime crashes or silent failures.

A safer approach is to cast defensively and fail fast if the response does not match expectations:

```kotlin
val data = result.data as? Map<*, *>
    ?: throw IllegalStateException("Invalid response format")

val serverTimeMillis = (data["timestamp"] as? Number)?.toLong()
    ?: throw IllegalStateException("Missing timestamp")
```

This ensures that parsing errors are immediately visible and prevents subtle bugs caused by unexpected response formats.